core/integration/odc/plugin.go

```diff
@@ -248,44 +250,97 @@ func (p *Plugin) queryPartitionStatus() {
-					WithField("oldState", existingPartition.State).
-					WithField("state", partitionInfo.State).
-					Info("ODC Partition state changed")
-
-				payload := map[string]interface{}{
-					"oldState":         existingPartition.State,
-					"newState":         partitionInfo.State,
-					"partitionId":      partitionInfo.PartitionId.String(),
-					"ddsSessionId":     partitionInfo.DdsSessionId,
-					"ddsSessionStatus": partitionInfo.DdsSessionStatus,
-				}
-				payloadJson, _ := json.Marshal(payload)
-
-				the.EventWriterWithTopic(TOPIC).WriteEvent(&pb.Ev_IntegratedServiceEvent{
-					Name:          "odc.partitionStateChanged",
-					EnvironmentId: id.String(),
-					Payload:       string(payloadJson[:]),
-				})
-
-				envMan := environment.ManagerInstance()
-				if envMan != nil {
-					go envMan.NotifyIntegratedServiceEvent(&event.OdcPartitionStateChangeEvent{
-						IntegratedServiceEventBase: common_event.IntegratedServiceEventBase{ServiceName: "ODC"},
-						EnvironmentId:              id,
-						State:                      partitionInfo.State,
+
+			// do we have the given partition on record already? if not, no state change detection is possible
+			if existingPartition, ok := p.cachedStatus.Partitions[id]; ok {
+
+				// detection of device state change + event publication
+				for deviceId, device := range partitionInfo.Devices {
+					oldDevice, hasDevice := existingPartition.Devices[deviceId]
+
+					oldState := "" // we presume the task didn't exist before
+
+					// if a device with this ID is already known to us from before
+					if hasDevice {
+						// if device state has changed
+						if oldDevice.State != device.State {
+							// if the state has changed, we take note of the previous state
+							oldState = oldDevice.State
+						} else {
+							// if the state hasn't changed, we bail
+							continue
+						}
+					}
+
+					device.EcsState = fairmq.ToEcsState(device.State, oldState)
+
+					// since the odc-state of the task has changed, we must publish the event
+					payload := map[string]interface{}{
+						"oldState":         oldState,
+						"newState":         device.State,
+						"ecsState":         device.EcsState,
+						"taskId":           device.TaskId,
+						"path":             device.Path,
+						"ignored":          device.Ignored,
+						"host":             device.Host,
+						"partitionId":      partitionInfo.PartitionId.String(),
+						"ddsSessionId":     partitionInfo.DdsSessionId,
+						"ddsSessionStatus": partitionInfo.DdsSessionStatus,
+					}
+					payloadJson, _ := json.Marshal(payload)
+
+					the.EventWriterWithTopic(TOPIC).WriteEvent(&pb.Ev_IntegratedServiceEvent{
+						Name:          "odc.deviceStateChanged",
+						EnvironmentId: id.String(),
+						Payload:       string(payloadJson[:]),
 					})
-				} else {
+				}
+
+				// detection of env (ODC partition) state change + event publication
+				if existingPartition.State != partitionInfo.State {
```

Comment:
```suggestion
				if oldPartitionInfo.State != partitionInfo.State {
```
could you use `old` for both device info and partition info? otherwise it's quite confusing to read if for some things it's `existing` and for others `old`.

It can be also `existing` for both.